contracts/governance/Vesting/OriginInvestorsClaim.sol

```diff
@@ -0,0 +1,150 @@
+
+	///@notice constant used for computing the vesting dates
+	uint256 public constant SOV_VESTING_CLIFF = 6 weeks;
+
+	VestingRegistry public vestingRegistry;
+	Staking public staking;
+	uint256 public kickoffTS;
+	IERC20 public SOVToken;
+	uint256 public vestingTerm;
+	bool public investorsListInitialized;
+
+	//user => flag whether user has admin role
+	mapping(address => bool) public admins;
+
+	// origin investors entitled to claim SOV
+	mapping(address => uint256) public investorsAmountsList;
+
+	event AdminAdded(address admin);
+	event AdminRemoved(address admin);
+	event InvestorsAmountsListSet(uint256 qty, uint256 totalAmount);
+	event ClaimVested(address indexed investor, address indexed receiver, uint256 amount);
+	event ClaimTransferred(address indexed investor, uint256 amount);
+
+	/**
+	 * @dev Throws if called by any account other than the owner or admin.
+	 */
+	modifier onlyAuthorized() {
+		require(isOwner() || admins[msg.sender], "OriginInvestorsClaim::onlyAuthorized: unauthorized");
+		_;
+	}
+
+	modifier onlyWhitelisted() {
+		require(investorsAmountsList[msg.sender] != 0, "OriginInvestorsClaim::onlyWhitelisted: not whitelisted or already claimed");
+		_;
+	}
+
+	constructor(address vestingRegistryAddress) public {
+		vestingRegistry = VestingRegistry(vestingRegistryAddress);
+		staking = Staking(vestingRegistry.staking());
+		kickoffTS = staking.kickoffTS();
+		SOVToken = staking.SOVToken();
+		vestingTerm = kickoffTS + SOV_VESTING_CLIFF;
+	}
+
+	function addAdmin(address _admin) public onlyOwner {
+		admins[_admin] = true;
+		emit AdminAdded(_admin);
+	}
+
+	function removeAdmin(address _admin) public onlyOwner {
+		admins[_admin] = false;
+		emit AdminRemoved(_admin);
+	}
+
+	//in case we have unclaimed tokens
+	function authorizedBalanceWithdraw(address toAddress) public onlyAuthorized {
+		require(
+			SOVToken.transfer(toAddress, SOVToken.balanceOf(address(this))),
+			"OriginInvestorsClaim::authorizedTransferBalance: transfer failed"
+		);
+	}
+
+	/**
+	 *  @notice the contract should be approved or transferred necessary amount of SOV prior to calling the function
+	 *	@param allowanceFrom address that approved total amount for this contract or 0x0 if transferred
+	 *  @param investors is the list of investors addresses
+	 *  @param claimAmounts is the list of amounts for investors investors[i] will receive claimAmounts[i] of SOV
+	 */
+	function setInvestorsAmountsList(
+		address[] memory investors,
+		uint256[] memory claimAmounts,
+		address allowanceFrom
+	) public onlyAuthorized {
+		uint256 totalAmount;
+		require(
+			investors.length == claimAmounts.length,
+			"OriginInvestorsClaim::setInvestorsAmountsList: investors.length != claimAmounts.length"
+		);
+		require(!investorsListInitialized, "OriginInvestorsClaim::setInvestorsAmountsList: the investors list has already been set");
+		investorsListInitialized = true;
```

Comment:
this means we'll be able to call the function only once. We've got over 3000 genesis investors. I doubt we can set all of them in a single transaction considering the RSK gas limit is only 6.8M.

Reply:
I looked up the numbers. we've got 3284 investors.

Reply:
how many can we process in a single tx?

Reply:
it takes one `sstore` per investor (20k gas), which translates to about 300 investors per transaction